### PR TITLE
MM-20312: Releasing Updated Plugins

### PR DESCRIPTION
**Pull request by lieut-data**
Relocate and expand the existing documentation on the `/mb cutPlugin` command to document releasing updated plugins.

Fixes: https://mattermost.atlassian.net/browse/MM-20312

### CODE REVIEW

File: site/content/internal/plugin-release-process.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+/mb cutPlugin --tag v0.13.0 --repo mattermost/mattermost-plugin-github
+```
+
+Matterbuild will create the given tag pointing at the current `master`, triggering CircleCI to [automatically create a GitHub release](https://github.com/mattermost/circleci-orbs/blob/3fb37c7920037c857a9ed9bc1a4e31be20092cdd/plugin-ci/orb.yml#L111-L120). Once the assets are built and uploaded to the release, matterbuild will download and sign using our production plugin signing key.
+
+Note that only authorized users have access to matterbuild.
+
+### Update the release description
+
+This step is not yet documented, but typically involves using [goreleaser](https://github.com/goreleaser/goreleaser) to update the release description.
```
Comment (lieut-data) [inline site/content/internal/plugin-release-process.md]: @hanzei, to kickstart this, would it make sense to list the setup you have for automating this?

File: site/content/internal/plugin-release-process.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+-    "version": "0.12.0",
++    "version": "0.13.0",
+    "min_server_version": "5.12.0",
+    ...
+}
+```
+
+Apply the updated manifest into the server and webapp:
+```
+make manifest
```
Comment (hanzei) [inline site/content/internal/plugin-release-process.md]: That is the wrong command
```suggestion
make apply
```

File: site/content/internal/plugin-release-process.md (hunk 1 above)
Comment (hanzei) [reply site/content/internal/plugin-release-process.md]: I run `goreleaser release --skip-validate --rm-dist` and then copy the content of  `dist/CHANGELOG.md` into GitHub.

File: site/content/internal/plugin-release-process.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+### Update the release description
+
+This step is not yet documented, but typically involves using [goreleaser](https://github.com/goreleaser/goreleaser) to update the release description.
+
+Automating this is scheduled as [MM-19848](https://mattermost.atlassian.net/browse/MM-19848).
+
+### Update the Plugin Marketplace
+
+Submit a ticket to the [Toolkit Team](/internal/rd-teams/#toolkit-team) to list the plugin in the marketplace.
```
Comment (hanzei) [inline site/content/internal/plugin-release-process.md]: Are we talking about a JIRA ticket here?

**Comment by lieut-data** [comment]
Note that I'm presently unable to fix the build. The `v5/` changes require Go modules to resolve, which in turn require switching away from golang.org/x/tools/go/packages to https://godoc.org/golang.org/x/tools/go/packages, but the effort to do so was non-trivial. More mana required, but for now all developer documentation updates are likely blocked.

Will address comments above separately -- thanks @hanzei!

File: site/content/internal/plugin-release-process.md (hunk 3 above)
Comment (lieut-data) [reply site/content/internal/plugin-release-process.md]: Yes, that's the proposal right now -- clarifying text.

**Comment by lieut-data** [comment]
I have fixed the build. Do not ask how.

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,9 @@
+module developer-documentation
```
Comment (agnivade) [inline go.mod]: This should be github.com/mattermost/mattermost-developer-documentation to avoid problems for importers of this module.

File: cmd/plugin-manifest-docs/plugin-manifest-docs.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+		}
+	case *ast.MapType:
+		return &TypeDocs{
+			Type:        Dict,
+			ValueSchema: astTypeDocs(x.Value, typesByName, info),
+		}
+	case *ast.InterfaceType:
+		return &TypeDocs{
+			Type: Interface,
+			// ValueSchema: astTypeDocs(x.Value, typesByName, info),
```
Comment (mgdelacroix) [inline cmd/plugin-manifest-docs/plugin-manifest-docs.go]: nit: comment left behind

**Comment by lieut-data** [comment]
Thanks, @agnivade & @mgdelacroix -- your comments made me realize I didn't delete the /old/ files from which I took most of the effort. To be clear, I didn't write all this AST code this time around (though I have contributed to it in the past ;P).

I've pushed changes to address all the feedback above!

File: .editorconfig [hunk 6]
```diff
@@ -5,17 +5,16 @@ root = true
 [*]
 end_of_line = lf
 insert_final_newline = true
-trim_trailing_whitespace = true
 charset = utf-8
-
-[*.{md,css,html}]
```
Comment (hanzei) [inline .editorconfig]: We still have some `html` and `css` files. Why did you remove the rules for them?

File: cmd/plugin-godocs/plugin-godocs.go [hunk 7]
```diff
@@ -1,23 +1,22 @@
...
 	"go/types"
 	"log"
 	"os"
 	"regexp"
 	"sort"
+
+	"golang.org/x/tools/go/packages"
+
+	_ "github.com/mattermost/mattermost-server/v5/plugin"
+	"github.com/pkg/errors"
```
Comment (hanzei) [inline cmd/plugin-godocs/plugin-godocs.go]: Nit: Move this next to `"golang.org/x/tools/go/packages"`

File: go.mod [hunk 8]
```diff
@@ -0,0 +1,9 @@
+module mattermost-developer-documentation
+
+go 1.13
+
+require (
+	github.com/mattermost/mattermost-server/v5 v5.18.0-rc.test
```
Comment (hanzei) [inline go.mod]: :100:

File: go.sum [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+github.com/magiconair/properties v1.8.1/go.mod h1:PppfXfuXeibc/6YijjN8zIbojt8czPbwD3XqdrwzmxQ=
+github.com/mailru/easyjson v0.7.0/go.mod h1:KAzv3t3aY1NaHWoQz1+4F1ccyAH66Jk7yos7ldAVICs=
+github.com/marstr/guid v0.0.0-20170427235115-8bdf7d1a087c/go.mod h1:74gB1z2wpxxInTG6yaqA7KrtM0NZ+RbrcqDvYHefzho=
+github.com/mattermost/go-i18n v1.11.0 h1:1hLKqn/ZvhZ80OekjVPGYcCrBfMz+YxNNgqS+beL7zE=
+github.com/mattermost/go-i18n v1.11.0/go.mod h1:RyS7FDNQlzF1PsjbJWHRI35exqaKGSO9qD4iv8QjE34=
+github.com/mattermost/gorp v2.0.1-0.20190301154413-3b31e9a39d05+incompatible h1:FN4zK2wNig7MVVsOsGEZ+LeIq0gUcudn3LEGgbodMq8=
+github.com/mattermost/gorp v2.0.1-0.20190301154413-3b31e9a39d05+incompatible/go.mod h1:0kX1qa3DOpaPJyOdMLeo7TcBN0QmUszj9a/VygOhDe0=
+github.com/mattermost/ldap v3.0.4+incompatible h1:SOeNnz+JNR+foQ3yHkYqijb9MLPhXN2BZP/PdX23VDU=
+github.com/mattermost/ldap v3.0.4+incompatible/go.mod h1:b4reDCcGpBxJ4WX0f224KFY+OR0npin7or7EFpeIko4=
+github.com/mattermost/mattermost-server v5.11.1+incompatible h1:LPzKY0+2Tic/ik67qIg6VrydRCgxNXZQXOeaiJ2rMBY=
```
Comment (hanzei) [inline go.sum]: I think `go mod tidy` is needed

File: go.mod (hunk 8 above)
Comment (mgdelacroix) [reply go.mod]: Didn't catch this at first, but that tag don't exist anymore, so we should point the server dep to a specific commit instead.

File: .editorconfig (hunk 6 above)
Comment (lieut-data) [reply .editorconfig]: I admit I'm not sure what happened here. Reverting changes -- thanks!

File: go.mod (hunk 8 above)
Comment (lieut-data) [reply go.mod]: Fixed. The pseudo-version selected seems "wrong", but the commit hash results in the desired behaviour. I suspect this will get fixed once v5.18.0 actually gets released.

**Comment by lieut-data** [comment]
Thanks for the feeback, @hanzei and @mgdelacroix, and sorry for the churn on some of these items that I should have caught earlier. Let me know if you have any further thoughts :)
